## Melbourne Housing Data Cleaning Project

In [2]:
# for basic operations
import numpy as np
import pandas as pd

# for data visualizations
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
import warnings
from warnings import filterwarnings
filterwarnings("ignore")

In [6]:
# lets import a dataset to perform the operations
data = pd.read_csv('melbourne.csv')

# lets check the shape of the dataset
data.shape

(23547, 21)

In [7]:
# lets check the structure of the dataset
data.head(10)

Suburb             Address  Rooms Type      Price Method  SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS   Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S   Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S   Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB   Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP   Biggin   
5  Abbotsford    40 Federation La      3    h   850000.0     PI   Biggin   
6  Abbotsford         55a Park St      4    h  1600000.0     VB   Nelson   
7  Abbotsford        16 Maugie St      4    h        NaN     SN   Nelson   
8  Abbotsford        53 Turner St      2    h        NaN      S   Biggin   
9  Abbotsford        99 Turner St      2    h        NaN      S  Collins   

         Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  03-09-2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  03-12-2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  04-02-2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  04-02-2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  04-03-2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   
5  04-03-2017       2.5    3067.0  ...       2.0  1.0      94.0           NaN   
6  04-06-2016       2.5    3067.0  ...       1.0  2.0     120.0         142.0   
7  06-08-2016       2.5    3067.0  ...       2.0  2.0     400.0         220.0   
8  06-08-2016       2.5    3067.0  ...       1.0  2.0     201.0           NaN   
9  06-08-2016       2.5    3067.0  ...       2.0  1.0     202.0           NaN   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.8014    144.9958  Northern Metropolitan   
1        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
2     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
3        NaN        Yarra  -37.8114    145.0116  Northern Metropolitan   
4     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   
5        NaN        Yarra  -37.7969    144.9969  Northern Metropolitan   
6     2014.0        Yarra  -37.8072    144.9941  Northern Metropolitan   
7     2006.0        Yarra  -37.7965    144.9965  Northern Metropolitan   
8     1900.0        Yarra  -37.7995    144.9974  Northern Metropolitan   
9     1900.0        Yarra  -37.7996    144.9989  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  
5        4019.0  
6        4019.0  
7        4019.0  
8        4019.0  
9        4019.0  

[10 rows x 21 columns]

In [14]:
# lets check the basic information about the data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23547 entries, 0 to 23546
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         23547 non-null  object 
 1   Address        23547 non-null  object 
 2   Rooms          23547 non-null  int64  
 3   Type           23547 non-null  object 
 4   Price          18396 non-null  float64
 5   Method         23547 non-null  object 
 6   SellerG        23547 non-null  object 
 7   Date           23547 non-null  object 
 8   Distance       23546 non-null  float64
 9   Postcode       23546 non-null  float64
 10  Bedroom2       19066 non-null  float64
 11  Bathroom       19063 non-null  float64
 12  Car            18921 non-null  float64
 13  Landsize       17410 non-null  float64
 14  BuildingArea   10018 non-null  float64
 15  YearBuilt      11540 non-null  float64
 16  CouncilArea    15656 non-null  object 
 17  Lattitude      19243 non-null  float64
 18  Longti

In [11]:
# lets first check if there is any null value present in the data
data.isnull().sum().sort_values(ascending=False)

BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Landsize          6137
Price             5151
Car               4626
Bathroom          4484
Bedroom2          4481
Longtitude        4304
Lattitude         4304
Distance             1
Propertycount        1
Postcode             1
Regionname           1
Date                 0
SellerG              0
Method               0
Type                 0
Rooms                0
Address              0
Suburb               0
dtype: int64

In [15]:
#lets check how many rows are having all the null values
data.isnull().all(axis = 1).sum()

0

In [16]:
# lets check the percentage of missing values in the columns of the dataset

percentage_of_missing_values = round(100*(data.isnull().sum()/len(data.index)), 2)
print(percentage_of_missing_values)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.88
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2         19.03
Bathroom         19.04
Car              19.65
Landsize         26.06
BuildingArea     57.46
YearBuilt        50.99
CouncilArea      33.51
Lattitude        18.28
Longtitude       18.28
Regionname        0.00
Propertycount     0.00
dtype: float64


<p style="color: green;">It can be noticed very easily from the above results that there are many columns have huge percentage of data missing from the dataset. There are columns which are having ```18% to 26%``` of values missing from the data, and also there are some columns which are having ```33% to 56%``` of values missing from the data.

* What action should we take, should we remove them all, or impute.
    * We ```cannot remove the columns having 18% to 26% missing values```, so In this case we will have to impute the missing values using one of the methods.
        * It is very important to impute the missing values for these columns as If we remove these columns from the data having only 18 to 26% of the missing values then we will face huge data loss, In that case, We might lose some of the very important and relevant data for the analysis and results.
    
    * But, ```We can remove the columns having 33 to 56% of missing values```, so In this case will have to drop these columns from the data.
        * It is very important to remove these columns because if we impute these missing values then we can introduce a huge bias into the data, which will be bad for the analysis of the data, It might also return inappropriate and misleading results.

In [17]:
# removing the three columns having 33% to 56% of missing values in the dataset
data = data.drop('BuildingArea', axis=1)
data = data.drop('YearBuilt', axis=1)
data = data.drop('CouncilArea', axis=1)

# data = data.drop(['Building Area', 'YearBuilt', 'CouncilArea'], axis = 1)

# lets check the columns left after the deletion of above listed columns
data.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'Lattitude', 'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [18]:
# count the number of rows having > 5 missing values
# use len(df.index)
len(data[data.isnull().sum(axis=1) > 5].index)

4278

In [19]:
# retaining the rows having <= 5 NaNs
data = data[data.isnull().sum(axis=1) <= 5]

In [21]:
# looking at the summary again
round(100*(data.isnull().sum()/len(data.index)), 2)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.71
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2          1.05
Bathroom          1.07
Car               1.81
Landsize          9.65
Lattitude         0.13
Longtitude        0.13
Regionname        0.00
Propertycount     0.00
dtype: float64

* It can be seen that ```Price Column``` still has large number of missing values i.e., 21% of the data is missing. If we impute these values it will introduce heavy bias into the data and the results will be misleading so It is better to discard the rows where Price is having missing values.

In [22]:
# removing all the rows where there is a missing value in the Price Column
data = data[~np.isnan(data['Price'])]

In [23]:
# look at the summary again
round(100*(data.isnull().sum()/len(data.index)), 2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         1.05
Bathroom         1.07
Car              1.76
Landsize         9.83
Lattitude        0.15
Longtitude       0.15
Regionname       0.00
Propertycount    0.00
dtype: float64

* Now, we have Landsize, which is having high percentage of missing values, lets check the range of values present in the Landsize attribute of the data, so that we can impute the missing values present in the Landsize Attribute.

In [24]:
# lets check the description of values in the Landsize Column
data['Landsize'].describe()

count     13603.000000
mean        558.116371
std        3987.326586
min           0.000000
25%         176.500000
50%         440.000000
75%         651.000000
max      433014.000000
Name: Landsize, dtype: float64

* From the above description of the range of values present in the landsize columns supported by mean, standard deviatio, minimum, maximum, 25% percentile, 50% percentile, 75% percentile etc.

* We can see that the Minimum value is 0 and Maximum value is 433014, there is huge difference so both mean and median function would not work

* Also, there is a huge difference in the 25th, 50th, and 75th percentiles indicating that if we impute the values for Landsize column we will most probably introduce bias into the data.

* It is most appropriate to remove the rows where we find null values for Landsize.

In [25]:
# removing  all the rows where there is a null value in the landsize column
data = data[~np.isnan(data['Landsize'])]

# lets check the summary again
round(100*(data.isnull().sum()/len(data.index)), 2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         0.00
Bathroom         0.01
Car              0.46
Landsize         0.00
Lattitude        0.16
Longtitude       0.16
Regionname       0.00
Propertycount    0.00
dtype: float64

* We can see that now we have very low fraction of missing values in the columns Bathroom, Car, Lattitude, and Longitude
* Lets check the range values of Lattitude and Longitude.

In [27]:
# describing the values for lattitude and longitude
data[['Lattitude','Longtitude']].describe()

Lattitude    Longtitude
count  13581.000000  13581.000000
mean     -37.809204    144.995221
std        0.079257      0.103913
min      -38.182550    144.431810
25%      -37.856820    144.929600
50%      -37.802360    145.000100
75%      -37.756400    145.058320
max      -37.408530    145.526350

In [66]:
# as there is minute difference between the values of min, mean, max, 25th, 50th, and 75th percentile. 
# so we can use mean function to impute the values in lattitude and longtitude.

data['Lattitude'] = data['Lattitude'].fillna(data['Lattitude'].mean())
data['Longtitude'].fillna(data['Longtitude'].mean(), inplace = True)

In [67]:
# lets check the values present in the Bathroom Column

data['Bathroom'].value_counts()

1.0    7517
2.0    4987
3.0     921
4.0     106
0.0      34
5.0      28
6.0       5
8.0       2
7.0       2
Name: Bathroom, dtype: int64

In [68]:
# lets also check the values present in the Car Column

data['Car'].value_counts()

2.0     5606
1.0     5515
0.0     1026
3.0      748
4.0      507
5.0       63
6.0       54
8.0        9
7.0        8
10.0       3
9.0        1
Name: Car, dtype: int64

In [69]:
# we have to impute the values in Categorical Columns using the Mode function

data['Bathroom'].fillna(data['Bathroom'].mode()[0], inplace = True)
data['Car'].fillna(data['Car'].mode()[0], inplace = True)

In [70]:
# lets check the missing values present in the data

data.isnull().sum().sum()

0